uqcsbot/api.py

```diff
@@ -305,5 +326,130 @@ def _on_group_joined(self, evt):
     def _on_group_left(self, evt):
         self._on_channel_deleted(evt)
 
-class MembersWrapper(object):
-    pass
+
+class UsersWrapper(object):
+    def __init__(self, bot: 'UQCSBot') -> None:
+        self._bot = bot
+        self._empty_store()
+        self._lock = threading.RLock()
+        self._initialised: bool = False
+
+        self._bind_handlers()
+
+    def _empty_store(self):
+        self._users_by_id: Dict[str, User] = {}
+        self._users_by_display: defaultdict[str, List[User]] = defaultdict(list)
+
+    def _add_user(self, data: dict):
+        user = User.from_dict(data)
+        self._users_by_id[user.user_id] = user
+        self._users_by_display[user.display_name].append(user)
+        return user
+
+    def _initialise(self):
+        with self._lock:
+            if self._initialised:
+                # Fast exit for initialisation while waiting on lock
+                return
+            self._empty_store()
+            for page in self._bot.api.users.list.paginate():
+                for user in page['users']:
+                    self._add_user(user)
+
+    def reload(self):
+        self._initialised = False
+        self._initialise()
+
+    def get(self, display_name_or_id: str, default: T = None, use_cache: bool =True) -> Union['User', List['User'], T]:
+        if use_cache and not self._initialised:
+            self._initialise()
+        if display_name_or_id in self._users_by_id:
+            return self._users_by_id[display_name_or_id]
+        elif display_name_or_id in self._users_by_display:
+            return self._users_by_display[display_name_or_id]
+        elif not use_cache:
+            resp = self._bot.api.users.info(user=display_name_or_id)
+            if not resp.get('ok'):
+                return default
+            with self._lock:
+                return self._add_user(resp['user'])
+        else:
+            return default
+
+    def populate_from_team_state(self, data_dict: dict):
+        with self._lock:
+            self._initialised = True
+            self._empty_store()
+            for user in data_dict['users']:
+                self._add_user(user)
+            self._bot.logger.info(f"Loaded {len(self._users_by_id)} users from team state")
+
+    def _bind_handlers(self) -> None:
+        PREFIX = "_on_"
+        for name in dir(type(self)):
+            if not name.startswith(PREFIX):
+                continue
+            attr = getattr(self, name)
+            mtype = name[len(PREFIX):]
+            self._bot.on(mtype, attr)
+
+    def _on_user_change(self, evt):
+        user = self._users_by_id[evt['user']['id']]
+        old_display_name = user.display_name
+        with self._lock:
+            user.update_from_dict(evt['user'])
+            if user.display_name != old_display_name:
+                self._users_by_display[old_display_name].remove(user)
+                if len(self._users_by_display[old_display_name]) == 0:
```

Comment:
can you add a comment here explaining code intent because its not immediately obvious